### ISSUE TITLE
Port of the collections module

### ISSUE THREAD

**Issue by JdeH** [QQuick/Transcrypt#442]
As multiple people have expressed an interest in porting the CPython `collections` module to Transcrypt, this issue has been openened to discuss its design.

**Comment by JdeH** [comment]
My five cents, as a starter:

With a module like `collections`, performance is important.
So while the API should match the API  of the CPython module as much as possible,
probably on the inside in some places native JS should be used.

My proposal is to make it work starting with JS 6, so to drop JS 5 compatibility for this module. (The closure compiler can always generate a polyfill). In this way certain optimizations of JS6 (like maybe `for...of` can be used to optimize performance.

**Comment by harto** [comment]
@JdeH sounds reasonable to me. Would it be wise to start with a pure Python implementation and selectively swap out parts for an ES6 implementation after profiling it?

A couple of other things I wondered about:
 - The CPython implementation relies on weakrefs for some of the collection types. Since there's no notion of weakrefs in JS, I wasn't sure if that was something that needed to be considered.
 - In general, how do you start on porting a module like this? Is the PyPy implementation a good place to start?

**Comment by JdeH** [comment]
I'll try to answer your questions as thorough as possible, but the answers I give are not absolute. It will take some figuring out to find the ideal route.

- In general with optimizations it is the case that one can best start out simple, to get a clean design, and then only optimize the bottlenecks. So that would be an argument in favor of starting with a pure Python implementation and then optimizing some parts directly in JS. 

- On the other hand some datastructures in Transcrypt currently are implemented with JS lib compatibility in mind, like e.g. dict, that's in fact a pure JS object. However JS6 has some new datatypes like Set and Map, that may be suitable as a basis for some types in the collections module, or it may not, I have no definitive opinion on that, it's just something that requires careful consideration.

- With regard to the weakref's:  Indeed in JS we'll have to do without them. This means that certain references have to be destroyed explicitly, since otherways objects will be kept alive for no reason and memoryleaks will result.

- With regard to PyPy: While inspiration may be found here, the PyPy implementation uses e.g. the threads module that isn't available in Transcrypt because of browser limitations. So I don't know if it will help much.

Maybe it's good to also look at some pure JS implementations, e.g. of a deque, as they may be quite efficient.

In general I expect that designing something like this will require some trial and error, refactoring and backtracking. Once it's there I expect it to be heavily used, so it pays of if indeed it's efficient.

Yet another point of consideration (though a dull subject) is licensing. The license of whatever code is borrowed from, should be compatible with Apache 2. MIT is, LGPL may be, GPL certainly isn't. The most important point is that Transcrypt is also allowed to be used in commercial closed source projects. One isn't obligated to publicize any sourcecode. Any 3rd party code used should grant the same freedom.

In general: If you embark on this (not any obligation), take your time. There's no hurry, quality is preferred over quantity here. And it really would be a very useful contributing that's going to benefit many!

**Comment by harto** [comment]
Thanks @JdeH! I'm interested in contributing. My approach will probably be to try and implement one of the collection types to start with, and I'll open a PR to get some early-stage feedback on the general approach. I hope to make a start this week.

**Comment by JdeH** [comment]
@harto Looking forward to your PR

**Comment by neriusmika** [comment]
In my fork I made OrderedDict, @JdeH what do you think about implementation?

**Comment by JdeH** [comment]
@neriusmika 

I've studied your` OrderedDict` implementation and I see that it's based on bare JavaScript objects, just as is the ordinary `dict`.

This approach was choosen for `dict` because I wanted it to be compatible with JSON literals, to be able to seamlessly use JS libraries. This led to limitations like keys always having to be strings, since actually they are names of attributes. To be able to use numerical keys, whenever a string key can be interpreted as a number, it is indeed interpreted as such. While this approach isn't ideal, it was the only way to get good interoperability with JavaScipt libraries.

With `OrdereredDict` there may be a somewhat different approach

1.  Since there are no OrderedDict literals of the form {key1: value1, key2, value2 ...} the JSON compatibility doesn't play a role here.

2. Another development is that Ecmascript 6 officially supports the Map datatype. It doesn't have the restriction of requiring string keys, so a true distinction can be made between 3 and '3' as key. Moreover the docs of Map say:

`A Map object iterates its elements in insertion order
`

[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)

And taken from that site:


> Objects are similar to Maps in that both let you set keys to values, retrieve those values, delete keys, and detect whether something is stored at a key. Because of this (and because there were no built-in alternatives), Objects have been used as Maps historically; however, there are important differences that make using a Map preferable in certain cases:
> 
> - The keys of an Object are Strings and Symbols, whereas they can be any value for a Map, including functions, objects, and any primitive.
> - You can get the size of a Map easily with the size property, while the number of properties in an Object must be determined manually.
> - A Map is an iterable and can thus be directly iterated, whereas iterating over an Object requires obtaining its keys in some fashion and iterating over them.
> - An Object has a prototype, so there are default keys in the map that could collide with your keys if you're not careful. As of ES5 this can be bypassed by using map = Object.create(null), but this is seldom done.
> - A Map may perform better in scenarios involving frequent addition and removal of key pairs.


Because of the points above, Map seems the ideal candidate as basis for an OrderedDict.
Extra "Python-style" methods could be added to it just the way you do that rightnow.
So it seems much of your code could be retained, but we'd have a real dict with typed keys rather than strings that are sometimes cleverly interpreted as numbers.

Since Transcrypt uses Google Closure, the minified result would still run on JS5.

What do you think?

**Comment by neriusmika** [comment]
I looked at Map type, but as I know, it doesn't support index like assignment like `a['b'] = 5`, it has methods get, set, so how can we predict at compile time, that object is OrderedDict and convert to `a.set('b', 5)` ?

**Comment by neriusmika** [comment]
But actually with Proxy I think it can be made. Still we can't use it with ES5, because there is no fully functional polyfill for proxies and never be. To make collections work with ES5 is almost impossible, without rewriting compiler and slowing down all the other types.

**Comment by JdeH** [comment]
Operator overloading would make this possible. We cannot predict that o is an OrderedDict, but at encountering o['b'] = 5, the operator overloading mechanism would do the trick if I'm right. This would require `__pragma__ ('opov')`, as currently is the case with concatenating lists with +, and cost some performance, though.

**Comment by JdeH** [comment]
I could live without ES5 compatibility. It's rapidly becoming a thing of the past. I don't think we should let Transcrypt's design be influenced to much by that.

@harto Since you're also planned to work on `collections`, any opinions on this?
Should there be some coordination to prevent duplication of effort?

**Comment by neriusmika** [comment]
I don't know, for me to use operator overloading for almost any operations with this type is huge usability issue (ant probably performance too). And inside this opov we can't use any normal index operations on the other types like dict, list, tuple, string.

**Comment by neriusmika** [comment]
We must forget about ES5 in collections. Now we need to focus just on implementation that would work with ES6. So if we will use opov, then instead of writing `od['a'] = d['b']`, we should write `c = d['b']` then opov, then `od['a'] = c`. I think it's not very elegant solution. And porting some third party library would be a nightmare. Or maybe I don't understand something?

**Comment by JdeH** [comment]
The normal index operations can be used, if I'm right, since they always serve as a fallback. But their performance would be lower.

About the use of opov, it wouldn't have to be so local!
If I am correct, 

opov
... (other statements)
od['a'] = d ['b']
... (other statements)
noopov

would just work, only not with optimum performance.
(I'll do a little experiment to find out if this indeed works, but it should)

3rd party libraries would work by adding just a few pragma's since they could cover many statements (or even just one).

Indeed lets forget ES5, agreed.
I am just bringing up some counterarguments here, not sure what should weigh in most.

**Comment by neriusmika** [comment]
If index operations work with Map type, then it's not a problem, but all examples that I saw was with get set methods. And if they don't work, then your example with opov will not work, because we can't distinguish ordereddict from normal dict at compile time.

**Comment by JdeH** [comment]
It isn't need to distinguish compile time. Runtime will do. If opov is on and no overloaded operator is present, it will default to native [].
But as said, it will cost some performance.

**Comment by neriusmika** [comment]
Jacques, I checked on JSFiddle and seems that Map supports map['a'] = 'abc'. So it can be done without opov.

**Comment by JdeH** [comment]
Well, that isn't  in the Mozilla docs...
Quite an omission.
Would be great, though!

**Comment by JdeH** [comment]
Tough luck, map ['a'] just adds an attribute...

```
m = new Map;

m [1] = 'one number'

m ['1'] = 'one string'

writeln (m [1])
writeln (m ['1'])

```
just prints

```
one string
one string

```

**Comment by JdeH** [comment]
Quite confusing if OrderedDicts do work without operator overloading, only completely different...

**Comment by neriusmika** [comment]
Indeed it's just an attribute. So you are the boss here, and you must decide, but for me opov just for every single assignment looks quite ugly.

**Comment by JdeH** [comment]
Lets not decide too quickly then!
As said, you don't need opov for every single assigment  (see example code).
Only one at the start of the module would do.
Just a performance hit, but on the other hand no extra list (although that may be under the hood in Map)...

**Comment by JdeH** [comment]
I have to attend to some private matters now, lets get back to this exploration.
The purpose is to find the best solution...

**Comment by neriusmika** [comment]
Yes I know, but opov on the whole module is not for real life applications. Actually I'm counting every millisecond in my code. With global opovs there is significant performance problems, and with opovs my code already looks like spaghetti. Until now I did not have much problems with non string keys in dictionaries, so I'll wait for other implementations :)

**Comment by neriusmika** [comment]
Jacques, what if we use my code with Proxy, but with Map instead plain Object and we intercept \_\_getattr\_\_, \_\_setattr\_\_, and redirect to Proxy target (actual Map type) get and set methods? I think that can work without opov. Of course, we must think what to do with functions like \_\_in\_\_ etc. but at least it can be solvable somehow.

**Comment by neriusmika** [comment]
@JdeH, I rewrote OrderedDict with Map and without opovs. Implementation is quite ugly, I made it just for review. Need to rest for better ideas.

**Comment by JdeH** [comment]
You've been working late!

I think this combination of proxy and map is indeed the best approach.
I like this idea very much, since it would avoid both the performance drop of operator overloading
and the limitations of js objects.

About the implementation, as you indicate it's a proof of concept, which I consider successful.
In a cleaned up version, probably `OrderedDict` could use the code code in `__core__.mod.js`:

```
// Mother of all classes
var object = {
    ...
}
```

rather than duplicating it.

The same holds for the code in __builtin__.mod.js:

```
function dict (objectOrPairs) {
    ....
}
```

although this may require a slight refactoring of the `dict` constructor, splitting into a more general function and a more specific one, doing things that only hold for either `dict` or `OrderedDict`, like creating instance to be either {} or a Map.

While the 'more general function' could be named something like `__dicty __` (analogous to `truthy(ness),` so it wouldn't pollute the namespace, it would be an "unofficial part" of the API, meaning that of course it couldn't be changed without breaking `OrderedDict`. Given the relatively large amount of code it would save, I consider that acceptable.

Apart from keeping the code smaller, it would result in less almost identical code to maintain.
What's your opinion on this?

**Comment by neriusmika** [comment]
Jacques, yes I was thinking about how to remove code duplication and that now there is no reason to make OrderedDict with pure javascript. With python encapsulating a Map, after compilation, only object construction would be different, but later we can inherit from that class. Because we aren't augmenting javascript objects with OrderedDict, I don't know which way is better.

**Comment by JdeH** [comment]
I also find it hard to predict which way works out better, pure JS or Python.
Some design goals should be:

- Since it's used so pervasively in interop with JS libs, `dict` should as lean and fast as possible (although an extra call in constructing it would be acceptable in my view)
- It is indeed desirable OrderedDict to share construction code, both with `__builtin__.dict` and with __core__.object

**Comment by neriusmika** [comment]
There is no hurry, probably I can make another variation and check usability and performance, then we can decide which one of them is better. Maybe even some frankenstein from both worlds. And I think now it would be better to implement namedtuple and counter. After that it would be clearer what we really need, because they are somehow similar to dictionaries.

**Comment by JdeH** [comment]
Something quite different, in between:

I am very impressed with the screendumps you sent my from the application you wrote for your job. It's exactly the thing that I hoped would happen, people using Python rather than JavaScript for large, complex in-browser applications. I am very curious about your experiences with things like maintainability and flexibility in the long run. I would expect Python to top JS in this respect.

**Comment by neriusmika** [comment]
Yes, first of all I tried react, vue, but they are completely unpythonic. Then I decided make from scratch with ideas from flask, the same style of declarative routing, blueprints, etc. Actually this application is working very well, fast, easy to use, but I'm making it in my spare time so there are still many incomplete things. But overall I like how it looks now.

**Comment by JdeH** [comment]
About counter and namedtuple: indeed, they should be drawn into the equation, although overseeing them all at once isn't easy. Indeed I think some experimenting is needed, and probably also some backtracking as collection types get added.

I completely agree that there's no hurry. Having good collection classes is at the core of having a truly usable language. So if you're willing to experiment and find your way to a near-optimum solution that would be valuable to many.

About the time scale of it all I've been thinking the following:

- I don't expect CPython to run in the browser with acceptable page load speeds,  unless browser manufactures would at least built in a Python VM, which they likely won't.
- No matter how many features are tacked on to JS, it will always remain a monument of organic growth, rather than design.

So the more I think about, the more I realize that Transcrypt probably has a long and happy life before it.

**Comment by neriusmika** [comment]
You know, I have one link about that in my bookmarks: http://live.julik.nl/2013/05/javascript-is-shit :)

**Comment by JdeH** [comment]
I've posted a little commercial there...

**Comment by neriusmika** [comment]
@JdeH I've made pure python implementation ant it was quite easy, the same test passed well, so now I need to benchmark those two implementations.

**Comment by JdeH** [comment]
Wauw!!!

I like the cleanness and simplicity of this implementation VERY much!
Would this be complete, including the constructor?
If that is the case I greatly prefer this over any JS implementation, even the expensive of some performance.

Am I right that the statement:

`key = list (self._map.js_keys ()) [idx]`

only does a conversion of the keys list, not a copy?

I can hardly believe it's this elegant...
Curious about the benchmark.
I  hope it's favorable, and if not, maybe some peephole optimizing using `__pragma__ ('js', ...)` does the trick.

**Comment by neriusmika** [comment]
Yes I know, but I looked how they implement this in javascript and they use `key = Array.from(map.keys())[idx]`, what is basically the same. I will think about that later. Maps just don't have indices on entries. After looking at generated javascript, there is almost no difference, between those implementations, but anyway I will test them.

**Comment by JdeH** [comment]
Incredible...

**Comment by neriusmika** [comment]
Made some benchmarking, results with 50000 randomly generated keys and values:

d assign: 9.785888671875ms
od1 assign: 38.93896484375ms
od2 assign: 63.10791015625ms
d iterate: 94.320068359375ms
od1 iterate: 76.48095703125ms
od2 iterate: 82.18310546875ms

where d - dict, od1 - ordereddict (javascript), od2 - ordereddict (python). The main problem why python version is slower are __get__ methods, and we can't do much here. I don't know is there a way to optimize them. Tested with -f flag, but anyway there is func.apply call and arguments slicing.

**Comment by neriusmika** [comment]
So the main performance bottleneck is func.apply (https://stackoverflow.com/questions/23769556/why-is-call-so-much-faster-than-apply)

**Comment by JdeH** [comment]
What does the assign do? Assigning what to what? A populated OD to another OD?

**Comment by JdeH** [comment]
Just a wild idea: could this extremely dirty trick help?

```
class A:
    def __init__ (self):
        self.x = 3
    
def f (self):
    return this.x
    
A.f = f

a = A ()

print (a.f ())

class B (A):
    pass
    
b = B ()

print (b.f ())

```

**Comment by JdeH** [comment]
Or, slightly less globally dirty:

```
class A:
    def __init__ (self, x):
        self.x = x
        
        def f (param):
            return this.x * param
            
        self.f = f
        
a = A (2)
a2 = A (3)

print (a.f (4), a2.f (5))

class B (A):
    pass
    
b = B (4)
b2 = B (5)

print (b.f (6), b2.f (7))

# Will print
# 8 15
# 24 35
``` 


Translates to:

```
	(function () {
		var __name__ = '__main__';
		var A = __class__ ('A', [object], {
			__module__: __name__,
			get __init__ () {return __get__ (this, function (self, x) {
				self.x = x;
				var f = function (param) {
					return this.x * param;
				};
				self.f = f;
			});}
		});
		var a = A (2);
		var a2 = A (3);
		print (a.f (4), a2.f (5));
		var B = __class__ ('B', [A], {
			__module__: __name__,
		});
		var b = B (4);
		var b2 = B (5);
		print (b.f (6), b2.f (7));
```

**Comment by neriusmika** [comment]
Thanks, Jacques, I'll try these solutions. Assignment in my test case just fills dictionaries from prepared arrays with keys and values:

```python
        for i in range(iterations):
            od1[keys[i]] = vals[i]
```
because in this loop every assignment calls \_\_setattr\_\_ via func.apply, that slowdowns the code, compared to javascript implementation.

**Comment by neriusmika** [comment]
By the way I remember from earlier benchmarking of my code, that \_\_get\_\_ always was the main bottleneck in loops.

**Comment by JdeH** [comment]
Alright, I've got it almost working: `__pragma__ ('jscalls', ...)`

**Comment by neriusmika** [comment]
Tried with the second example (f) and it performed even better than javascript implementation!

**Comment by neriusmika** [comment]
Run several times and it's about 5-10% faster than js version. Now I have another question - why? :rofl:

**Comment by JdeH** [comment]
Good question...

**Comment by JdeH** [comment]
I've commit the -jc / --jscall switch and pragma.
Curious what it does for your benchmarks.

**Comment by neriusmika** [comment]
Thanks, I'll report after testing.

**Comment by JdeH** [comment]
This:

```
class A:
    def __init__ (self, x):
        self.x = x
                
    __pragma__ ('jscall')

    def f (self, param):
        return self.x * param
        
    __pragma__ ('nojscall')
    
    def g (self, param):
        return self.x * param

a = A (2)
a2 = A (3)

print (a.f (4), a2.f (5))

class B (A):
    pass
    
b = B (4)
b2 = B (5)

print (b.f (6), b2.f (7))

# Will print
# 8 15
# 24 35

```

now generates this:

```
		var A = __class__ ('A', [object], {
			__module__: __name__,
			get __init__ () {return __get__ (this, function (self, x) {
				self.x = x;
			});},
			f: function (param) {
				var self = this;
				return self.x * param;
			},
			get g () {return __get__ (this, function (self, param) {
				return self.x * param;
			});}
		});
		var a = A (2);
		var a2 = A (3);
		print (a.f (4), a2.f (5));
		var B = __class__ ('B', [A], {
			__module__: __name__,
		});
		var b = B (4);
		var b2 = B (5);
		print (b.f (6), b2.f (7));

```

Only thing I'd like to get rid of is:

`var self = this`

It could be done once on class level, e.g. by metaclass, or by some kind of alias.
Don't know yet.
May hinder performance a bit but is curable, I think.

**Comment by neriusmika** [comment]
I got runtime error "Uncaught SyntaxError: Unexpected token }" (compiled with -js switch):
```javascript
var __Envir__ = __class__ ('__Envir__', [object], {
						__module__: __name__,
						__init__: function () {
							var self = this;
							self.interpreter_name = 'python';
							self.transpiler_name = 'transcrypt';
							self.transpiler_version = '3.6.85';
							self.target_subdir = '__javascript__';
						}, '__init__'
					});
```

**Comment by neriusmika** [comment]
And when switched of fastcall, then inside the same function "this" equals to null

**Comment by JdeH** [comment]
I ran the tests only with the local pragma.
Am afraid I cannot fix it rightnow.
Could you try that?

**Comment by neriusmika** [comment]
With pragmas, the code runs normally, so the global switch causes errors.

**Comment by JdeH** [comment]
Allright, will try to fix it tomorrow.

**Comment by neriusmika** [comment]
Thank you Jacques, it is very valuable fix, now performance is great.

**Comment by JdeH** [comment]
Good!
Then we can have Python-only collections, I guess?

**Comment by neriusmika** [comment]
Yes, I think. The code would be much cleaner.

**Comment by neriusmika** [comment]
Found the error with fastcalls and fixed it, tomorrow I'll commit to my fork. Just want to check other issues.

**Comment by neriusmika** [comment]
Commited some fixes:
* fastcalls now work correctly;
* simplified visitFunctionDef;
* with jscall the first parameter can have any name ("self" is not hardcoded);
* jscall works with classmethod, staticmethod and property decorators (except all other decorators, I don't know how to pass "self" to them, so temporary just disabled this feature);
* added "deleteProperty" for proxies, so \_\_delattr\_\_  is working now (just need to refactor this part of code, if class doesn't have \_\_setattr\_\_, \_\_getattr\_\_, \_\_delattr\_\_ probably there is no need to define handlers for them).

I did not check global -jc switch, because I'm tired now, maybe will look tomorrow. Also we need to solve "self" problem somehow when jscall enabled, the first line in method body is not very big problem, but we can't decorate these methods. And thanks for the jscall feature!

**Comment by JdeH** [comment]
Ziet er goed uit.

**Comment by neriusmika** [comment]
Bedankt. But actually I found some problems with jscall approach, not everywhere it can be applied. The shitty javascript "this" scope problem. Sometimes "this" is passed as null, later I'll try to make a test.

**Comment by neriusmika** [comment]
@JdeH, probably there is no any easy solution to this problem. I think we must leave only jscall pragma without global switch, because it would be to dangerous. And use that pragma on specific methods where performance is important. There is no need to fix decorators for them too. Even simple transcrypt dictionaries have this problem with "this":
```python
d1 = {'a': 'A', 'b': 'B'}
d2 = {'c': 'C', 'd': 'D'}
update = d2.update
update(d1)  <== TypeError: Cannot set property 'a' of undefined
```

**Comment by JdeH** [comment]
OK, no problem at all.
It's nice that we have this optimization where it really counts.
And quite some warning words about its limitations should be in the docs.

I will take the global switch out again (currently a bit thinly spread)
It would be no good anyhow, making bound pointer assignment impossible.

However the possibilty to write a fast container in Python as part of a possibly heavily used lib is a big win.

**Comment by neriusmika** [comment]
Yes, I have tested on my code, when you carefully select which methods to optimize, then there are no problems.

**Comment by harto** [comment]
This is awesome! I'm just getting started with `defaultdict`, which I think should be relatively simple. I'll be sure to report progress in here, so we can coordinate efforts.

@neriusmika out of interest, what does the `__bases__` assignment in `OrderedDict` accomplish?
```
class OrderedDict:
    __bases__ = [dict]
```

**Comment by neriusmika** [comment]
@harto, because we can't inherit from dict type, without this line OrderedDict will show that it's inherited from "object", so functions like isinstance, issubclass will behave differently comparing to cpython. Actually I don't remember exactly, but I think that OrderedDict is inherited from dict :)

**Comment by harto** [comment]
Got it! And is the reason we can't inherit from dict because it's implemented in JS? Or because OrderedDict doesn't need to inherit dict's implementation (because it uses Map internally)?

**Comment by neriusmika** [comment]
It's because Transcrypt dict is just augmented js object. Implementation is different from standard classes and it's impossible to inherit from dict. That's way we need to make these classes from scratch.

**Comment by harto** [comment]
I had some free time this evening to spend on `defaultdict`.

My understanding is that if I implemented it as a Python class, users would need to use `opov` to be able to do dict lookups. I don't know if that's expected or desirable. If not, it occurred to me that we might be able to use a JS `Proxy` to get the desired behaviour. Let me know if you have any thoughts about this.

**Comment by JdeH** [comment]
@harto: I know @neriusmika has met this dilemma n working on the OrderedDict class, so he's probably in the best position to answer this. I remember he has tried JS Proxy.

@neriusmika: I should again dive into this, but currently I don't remember exactly what were the considerations. I know you started out with JS / JS Proxy. I don't see the opov pragma in your committed test code but I do see `__setitem__` etc. in the collections code. Could you clarify to @harto (and refresh my memory on this)

KR
J

**Comment by neriusmika** [comment]
Jacques, there are some problems and now I can't decide which way to go. 

Using python source with proxies:
* bad performance (of course we need to decide what is acceptable performance, but it's slower than pure js objects, like current dict implementation).
* does not support non string keys (Proxies limitation).

Using pure js implementation:
* impossible to inherit from such pseudo classes.
* does not support non string keys.
* good performance.

The third possibility is to make with python classes and operator overloading, but every access to items must be overloaded, so again:
* extra call would slowdown performance.
* higher probability of mistakes.
* support any type as keys (with internal Map).

P.S. \_\_setitem\_\_ in my implementation is the method called by proxy handler.

**Comment by JdeH** [comment]
@harto: This sums it up nicely. By interpreting each string that denotes a number as a numerical key we can have numerical keys, just as with the "normal" sets/objects. It covers the majority of cases.

@neriusmika: I know you know this, but I mention it to clarify for @harto.


@harto @neriusmika 
For collection classes both time and space performance are important, although I slightly favor time performance. We absolutely need to keep up with JavaScript on that point, to offer an uncompromising alternative. Speed and compactness are the niche and unique selling point of Transcrypt, making it truly usable for professional applications.

If that requires pure JS I am fine with it. But in that case I expect it to be very important to share as much code as possible between the diverse collection classes in the collections module. So if you both work on it it would require cooperation.

**Comment by EvanCarroll** [comment]
Just wanted in to resurrect this. Can someone summarize the current problems that we're facing? My vote is to just use Map. The important thing is that the code works, not that it's Javascript-like. Map fits closest to the Python datastructure: it's ordered. The only drawback is the `.set` and `.get` problem? Is that an issue? If the Python datastructure is a dictionary, generate a Map and always use `set` and `get`?

**Comment by JdeH** [comment]
Indeed this has come to a standstill.
Currently I'm quite tied up in getting 3.7 out and other participants seem to have similar considerations.

Summarizing:
- It would be very useful to have a _collections_ module with (as much as possible) the same API as the one that comes with CPython 3.7
- Performance (speed and compact storage) is very important for this module, since it will probably be intensively used for large in-memory data structures.
- In my view JS5 compatibility can be dropped (and, if needed, achieved with the usual polyfills).
- The JS Map type is probably most suitable as a basis for the _OrderedDict_ class of _collections_.
- As Python _OrderedDict_ support [] (index like assignment), Transcrypt _OrderedDict_ should as well. This is not without (performance) problems, as can be seen from the above discussion.
- Implementing ANY dict as a Map isn't an option, since JTranscrypt (Python) dict literals like {1, 2, '4', [10, 20], True} should translate to plain JS objects to remain compatible with the countles JS libs that use object literals (JSON) as a way to initialize c.q. configure diverse library objects. So while OrderedDict probably should be based on Map, dict should not.
- To achieve code compactness, code could be shared between similar data structures, although speed and data compactness are more important.

The current holdup, to be exact, is making an evidence (benchmark) based choice from the possibilities summed up at:

https://github.com/QQuick/Transcrypt/issues/442#issuecomment-365253302

I propose to use the criteria described in:

https://github.com/QQuick/Transcrypt/issues/442#issuecomment-365259916

Guess that sums if up.
If you or anyone else would have available the time and skills to get this going again, it would be much appreciated.
Whatever is chosen will greatly influence performance of Transcrypt applications, so it's important to do "the right thing". Which isn't easy and probably the reason this got stuck for so long...

**Comment by EvanCarroll** [comment]
> Implementing ANY dict as a Map isn't an option, since Transcrypt (Python) dict literals like {1, 2, '4', [10, 20], True} should translate to plain JS objects **to remain compatible with the countles JS libs that use object literals (JSON) as a way to initialize c.q. configure diverse library objects.** So while OrderedDict probably should be based on Map, dict should not.

I am not following on this point? As far as I can see, the concern here is that we need to provide a constructor that builds Maps from Objects for integration into libraries that Transcrypt generates. The reality is that the Object in Javascript is a poor fit for what even Javascript programmers use it for, those should all be Maps (with very few exceptions). That Object-literal notation doesn't create Maps is a *huge* design-flaw in JavaScript only because it makes it easy to do the wrong thing. But, we shouldn't be putting in more work to do the wrong thing for third party libraries. If we're generating code that does what the Python code does and uses Maps internally, that's **great**. I'm confused at why we're holding up a direct implementation on Map. We can perhaps provide convenience here, but it seems like our stance should be

* Map is the JavaScript analog of a Python dict
* JavaScript libraries should use Map instead

I definitely would like to see this done. I have a huge project that I think is mostly held up by this. I just want to be sure I'm on the same page and not stepping on toes or going against the grain. Then I'll dive in. **BTW really cool project, I'm trying real hard not to go all fan boy on the issue tracker**
